quinn-proto/src/connection/assembler.rs

```diff
@@ -94,7 +94,19 @@ impl Assembler {
     // counter to the new number of chunks left in the heap so that we can decide
     // when to defragment the queue again if necessary.
     fn defragment(&mut self) {
-        let buffered = self.data.iter().map(|c| c.bytes.len()).sum::<usize>();
+        let high_utilization_over_allocation = self
+            .data
+            .iter()
+            .filter(|b| b.high_utilization())
+            .map(|b| b.over_allocation)
+            .sum::<usize>();
+        let include_highly_utilized = high_utilization_over_allocation > DEFRAGMENTATION_THRESHOLD;
```

Comment:
OK, I will remove it again. But do you think it would make sense to retain skipping defragmentation for a `Buffer` with `over_allocation == 0`?

Reply:
> OK, I will remove it again. 

Sorry if this is a bit churny; I don't mind waiting for @Matthias247 to weigh in on how smart we should be here, since I think he was advocating for that in the first place.

> But do you think it would make sense to retain skipping defragmentation for a Buffer with over_allocation == 0?

The original logic takes care to avoid defragmenting the same data multiples times; it might be worth preserving that somehow, perhaps even by just retaining the `defragmented` counter. For newly inserted data, there will always be some overhead due to headers.

> The defragmentation decision will then look like this:

That looks great! Just making some numbers up based on nothing, maybe 1.5 and 2^12?

Reply:
> The original logic takes care to avoid defragmenting the same data multiples times; it might be worth preserving that somehow, perhaps even by just retaining the `defragmented` counter. For newly inserted data, there will always be some overhead due to headers.

Are you sure that the original logic does that? To me it looks like the `defragmented` was necessary to know how many new inserts were done since the last defragmentation and on defragmentation all data was copied into a single new allocation.

Reply:
Oops, you're quite right, I was confused. In any case, it seems like skipping already-defragmented chunks might be relatively simple, so perhaps worth considering.

Reply:
The new version will skip already-defragmented chunks.